# Mathematical operations
addition, subtraction, division, multiplication, etc. One thing to note here is that NumPy comes with its own set of methods and operations. You should not assume that the operators will work in the same way as lists. Look at the example below.

In [ ]:
import numpy as np
# Create 2 lists with elements (1, 2, 3) and (4, 5, 6)
list1 = [1,2,3]
list2 = [4,5,6]

# Use the '+' operator and print the results
list1+list2


In [ ]:
import numpy as np

In [ ]:
# Convert the above lists to arrays and use the '+' operator or np.add() for element wise addition

np_1 = np.array([1,2,3])
np_2 = np.array([4,5,6])


print(np_1+np_2+np_1)
print()
print(np.add(np_1,np_2))

    Arithmatic operators/Functions
        np.add()            =>  +
        np.subtract()       =>  -
        np.multiply()       =>  *
        np.divide()         =>  /
        np.floor_divide()   =>  //
        np.mod()            =>  %
        np.power()

    7=3*2+1
    Dividend
    divisor
    quotient  => Floor 
    remainder => Mod

In [ ]:
arr_3= np.arange(5)
arr_3,arr_3+5,arr_3*5,arr_3-2

In [ ]:
lst=[i for i in range(5)]
lst+5

In [ ]:
# use the '*' operator or np.multiply() for element wise multiplication in below cell


In [ ]:
#answer here

In [ ]:
np_1,np_2

In [ ]:
# use the '/' operator or np.divide() for element wise division

print(np_1/np_2)
print()
print(np.divide(np_1,np_2))

In [ ]:
# use the '-' operator or np.subtract() for element wise subtraction

print(np_2-np_1)
print()
print(np.subtract(np_2,np_1))

In [ ]:
# use the '//' operator or np.floor_divide() for element wise floor division
print(np_2//np_1)
print()
print(np.floor_divide(np_2,np_1))

In [ ]:
# use the '%' operator or np.mod() for element wise modulo
print(np_2%np_1)
print()
print(np.mod(np_2,np_1))

In [ ]:
np_1,np_2

In [ ]:
# use the '**' operator or np.power() for element wise exponentiation
print(np_2**np_1)
print()
print(np.power(np_2,np_1))

In [ ]:
# Scaler multiplication
np_1 = np.array([1,2,3])*5
np_1

In [ ]:
# Scaler addition
np_1 = np.array([1,2,3])+5
np_1

In [ ]:
# To find Scaler division in below cell using numpy array


In [ ]:
#answer here

In [ ]:
np.array([10,20,30])-5

In [ ]:
# # Scaler subtraction
np_1 = np.array([10,20,30])-5
np_1

In [ ]:
# squaring the terms
print(np.power(np.array([1,2,-1,-2]),2))

In [ ]:
# np.sin(1.57079633)
print(np.cos(0))

In [ ]:
np.sin(45)

In [ ]:
# np.log()
print(np.log(10))

In [ ]:
# getting the absolute value of the elements
print(np.absolute(np.array([1,2,-1,-2])))

# Case Study: Cricket Tournament
A panel wants to select players for an upcoming league match based on their fitness. Players from all significant cricket clubs have participated in a practice match, and their data is collected. Let us now explore NumPy features using the player's data.

Height of the payers is stored as a regular Python list: height_in. The height is expressed in inches.
Weight of the payers is stored as a regular Python list: weight_lb. The weight is expressed in pounds.
Multiplication with a factor

Now, let's convert the height and weight of the players into more generic units.

Height in metres
Weight in kilograms

In [ ]:
heights_in = [74, 74, 72, 72, 73, 69, 69, 71, 76, 71, 73, 73, 74, 74, 69, 70, 73, 75, 78, 79, 76, 74, 76, 72, 71, 75, 77, 74, 73, 74, 78, 73, 75, 73, 75, 75, 74, 69, 71, 74, 73, 73, 76, 74, 74, 70, 72, 77, 74, 70, 73, 75, 76, 76, 78, 74, 74, 76, 77, 81, 78, 75, 77, 75, 76, 74, 72, 72, 75, 73, 73, 73, 70, 70, 70, 76, 68, 71, 72, 75, 75, 75, 75, 68, 74, 78, 71, 73, 76, 74, 74, 79, 75, 73, 76, 74, 74, 73, 72, 74, 73, 74, 72, 73, 69, 72, 73, 75, 75, 73, 72, 72, 76, 74, 72, 77, 74, 77, 75, 76, 80, 74, 74, 75, 78, 73, 73, 74, 75, 76, 71, 73, 74, 76, 76, 74, 73, 74, 70, 72, 73, 73, 73, 73, 71, 74, 74, 72, 74, 71, 74, 73, 75, 75, 79, 73, 75, 76, 74, 76, 78, 74, 76, 72, 74, 76, 74, 75, 78, 75, 72, 74, 72, 74, 70, 71, 70, 75, 71, 71, 73, 72, 71, 73, 72, 75, 74, 74, 75, 73, 77, 73, 76, 75, 74, 76, 75, 73, 71, 76, 75, 72, 71, 77, 73, 74, 71, 72, 74, 75, 73, 72, 75, 75, 74, 72, 74, 71, 70, 74, 77, 77, 75, 75, 78, 75, 76, 73, 75, 75, 79, 77, 76, 71, 75, 74, 69, 71, 76, 72, 72, 70, 72, 73, 71, 72, 71, 73, 72, 73, 74, 74, 72, 75, 74, 74, 77, 75, 73, 72, 71, 74, 77, 75, 75, 75, 78, 78, 74, 76, 78, 76, 70, 72, 80, 74, 74, 71, 70, 72, 71, 74, 71, 72, 71, 74, 69, 76, 75, 75, 76, 73, 76, 73, 77, 73, 72, 72, 77, 77, 71, 74, 74, 73, 78, 75, 73, 70, 74, 72, 73, 73, 75, 75, 74, 76, 73, 74, 75, 75, 72, 73, 73, 72, 74, 78, 76, 73, 74, 75, 70, 75, 71, 72, 78, 75, 73, 73, 71, 75, 77, 72, 69, 73, 74, 72, 70, 75, 70, 72, 72, 74, 73, 74, 76, 75, 80, 72, 75, 73, 74, 74, 73, 75, 75, 71, 73, 75, 74, 74, 72, 74, 74, 74, 73, 76, 75, 72, 73, 73, 73, 72, 72, 72, 72, 71, 75, 75, 74, 73, 75, 79, 74, 76, 73, 74, 74, 72, 74, 74, 75, 78, 74, 74, 74, 77, 70, 73, 74, 73, 71, 75, 71, 72, 77, 74, 70, 77, 73, 72, 76, 71, 76, 78, 75, 73, 78, 74, 79, 75, 76, 72, 75, 75, 70, 72, 70, 74, 71, 76, 73, 76, 71, 69, 72, 72, 69, 73, 69, 73, 74, 74, 72, 71, 72, 72, 76, 76, 76, 74, 76, 75, 71, 72, 71, 73, 75, 76, 75, 71, 75, 74, 72, 73, 73, 73, 73, 76, 72, 76, 73, 73, 73, 75, 75, 77, 73, 72, 75, 70, 74, 72, 80, 71, 71, 74, 74, 73, 75, 76, 73, 77, 72, 73, 77, 76, 71, 75, 73, 74, 77, 71, 72, 73, 69, 73, 70, 74, 76, 73, 73, 75, 73, 79, 74, 73, 74, 77, 75, 74, 73, 77, 73, 77, 74, 74, 73, 77, 74, 77, 75, 77, 75, 71, 74, 70, 79, 72, 72, 70, 74, 74, 72, 73, 72, 74, 74, 76, 82, 74, 74, 70, 73, 73, 74, 77, 72, 76, 73, 73, 72, 74, 74, 71, 72, 75, 74, 74, 77, 70, 71, 73, 76, 71, 75, 74, 72, 76, 79, 76, 73, 76, 78, 75, 76, 72, 72, 73, 73, 75, 71, 76, 70, 75, 74, 75, 73, 71, 71, 72, 73, 73, 72, 69, 73, 78, 71, 73, 75, 76, 70, 74, 77, 75, 79, 72, 77, 73, 75, 75, 75, 73, 73, 76, 77, 75, 70, 71, 71, 75, 74, 69, 70, 75, 72, 75, 73, 72, 72, 72, 76, 75, 74, 69, 73, 72, 72, 75, 77, 76, 80, 77, 76, 79, 71, 75, 73, 76, 77, 73, 76, 70, 75, 73, 75, 70, 69, 71, 72, 72, 73, 70, 70, 73, 76, 75, 72, 73, 79, 71, 72, 74, 74, 74, 72, 76, 76, 72, 72, 71, 72, 72, 70, 77, 74, 72, 76, 71, 76, 71, 73, 70, 73, 73, 72, 71, 71, 71, 72, 72, 74, 74, 74, 71, 72, 75, 72, 71, 72, 72, 72, 72, 74, 74, 77, 75, 73, 75, 73, 76, 72, 77, 75, 72, 71, 71, 75, 72, 73, 73, 71, 70, 75, 71, 76, 73, 68, 71, 72, 74, 77, 72, 76, 78, 81, 72, 73, 76, 72, 72, 74, 76, 73, 76, 75, 70, 71, 74, 72, 73, 76, 76, 73, 71, 68, 71, 71, 74, 77, 69, 72, 76, 75, 76, 75, 76, 72, 74, 76, 74, 72, 75, 78, 77, 70, 72, 79, 74, 71, 68, 77, 75, 71, 72, 70, 72, 72, 73, 72, 74, 72, 72, 75, 72, 73, 74, 72, 78, 75, 72, 74, 75, 75, 76, 74, 74, 73, 74, 71, 74, 75, 76, 74, 76, 76, 73, 75, 75, 74, 68, 72, 75, 71, 70, 72, 73, 72, 75, 74, 70, 76, 71, 82, 72, 73, 74, 71, 75, 77, 72, 74, 72, 73, 78, 77, 73, 73, 73, 73, 73, 76, 75, 70, 73, 72, 73, 75, 74, 73, 73, 76, 73, 75, 70, 77, 72, 77, 74, 75, 75, 75, 75, 72, 74, 71, 76, 71, 75, 76, 83, 75, 74, 76, 72, 72, 75, 75, 72, 77, 73, 72, 70, 74, 72, 74, 72, 71, 70, 71, 76, 74, 76, 74, 74, 74, 75, 75, 71, 71, 74, 77, 71, 74, 75, 77, 76, 74, 76, 72, 71, 72, 75, 73, 68, 72, 69, 73, 73, 75, 70, 70, 74, 75, 74, 74, 73, 74, 75, 77, 73, 74, 76, 74, 75, 73, 76, 78, 75, 73, 77, 74, 72, 74, 72, 71, 73, 75, 73, 67, 67, 76, 74, 73, 70, 75, 70, 72, 77, 79, 78, 74, 75, 75, 78, 76, 75, 69, 75, 72, 75, 73, 74, 75, 75, 73]


In [ ]:
weights_lb = [180, 215, 210, 210, 188, 176, 209, 200, 231, 180, 188, 180, 185, 160, 180, 185, 189, 185, 219, 230, 205, 230, 195, 180, 192, 225, 203, 195, 182, 188, 200, 180, 200, 200, 245, 240, 215, 185, 175, 199, 200, 215, 200, 205, 206, 186, 188, 220, 210, 195, 200, 200, 212, 224, 210, 205, 220, 195, 200, 260, 228, 270, 200, 210, 190, 220, 180, 205, 210, 220, 211, 200, 180, 190, 170, 230, 155, 185, 185, 200, 225, 225, 220, 160, 205, 235, 250, 210, 190, 160, 200, 205, 222, 195, 205, 220, 220, 170, 185, 195, 220, 230, 180, 220, 180, 180, 170, 210, 215, 200, 213, 180, 192, 235, 185, 235, 210, 222, 210, 230, 220, 180, 190, 200, 210, 194, 180, 190, 240, 200, 198, 200, 195, 210, 220, 190, 210, 225, 180, 185, 170, 185, 185, 180, 178, 175, 200, 204, 211, 190, 210, 190, 190, 185, 290, 175, 185, 200, 220, 170, 220, 190, 220, 205, 200, 250, 225, 215, 210, 215, 195, 200, 194, 220, 180, 180, 170, 195, 180, 170, 206, 205, 200, 225, 201, 225, 233, 180, 225, 180, 220, 180, 237, 215, 190, 235, 190, 180, 165, 195, 200, 190, 190, 185, 185, 205, 190, 205, 206, 220, 208, 170, 195, 210, 190, 211, 230, 170, 185, 185, 241, 225, 210, 175, 230, 200, 215, 198, 226, 278, 215, 230, 240, 184, 219, 170, 218, 190, 225, 220, 176, 190, 197, 204, 167, 180, 195, 220, 215, 185, 190, 205, 205, 200, 210, 215, 200, 205, 211, 190, 208, 200, 210, 232, 230, 210, 220, 210, 202, 212, 225, 170, 190, 200, 237, 220, 170, 193, 190, 150, 220, 200, 190, 185, 185, 200, 172, 220, 225, 190, 195, 219, 190, 197, 200, 195, 210, 177, 220, 235, 180, 195, 195, 190, 230, 190, 200, 190, 190, 200, 200, 184, 200, 180, 219, 187, 200, 220, 205, 190, 170, 160, 215, 175, 205, 200, 214, 200, 190, 180, 205, 220, 190, 215, 235, 191, 200, 181, 200, 210, 240, 185, 165, 190, 185, 175, 155, 210, 170, 175, 220, 210, 205, 200, 205, 195, 240, 150, 200, 215, 202, 200, 190, 205, 190, 160, 215, 185, 200, 190, 210, 185, 220, 190, 202, 205, 220, 175, 160, 190, 200, 229, 206, 220, 180, 195, 175, 188, 230, 190, 200, 190, 219, 235, 180, 180, 180, 200, 234, 185, 220, 223, 200, 210, 200, 210, 190, 177, 227, 180, 195, 199, 175, 185, 240, 210, 180, 194, 225, 180, 205, 193, 230, 230, 220, 200, 249, 190, 208, 245, 250, 160, 192, 220, 170, 197, 155, 190, 200, 220, 210, 228, 190, 160, 184, 180, 180, 200, 176, 160, 222, 211, 195, 200, 175, 206, 240, 185, 260, 185, 221, 205, 200, 170, 201, 205, 185, 205, 245, 220, 210, 220, 185, 175, 170, 180, 200, 210, 175, 220, 206, 180, 210, 195, 200, 200, 164, 180, 220, 195, 205, 170, 240, 210, 195, 200, 205, 192, 190, 170, 240, 200, 205, 175, 250, 220, 224, 210, 195, 180, 245, 175, 180, 215, 175, 180, 195, 230, 230, 205, 215, 195, 180, 205, 180, 190, 180, 190, 190, 220, 210, 255, 190, 230, 200, 205, 210, 225, 215, 220, 205, 200, 220, 197, 225, 187, 245, 185, 185, 175, 200, 180, 188, 225, 200, 210, 245, 213, 231, 165, 228, 210, 250, 191, 190, 200, 215, 254, 232, 180, 215, 220, 180, 200, 170, 195, 210, 200, 220, 165, 180, 200, 200, 170, 224, 220, 180, 198, 240, 239, 185, 210, 220, 200, 195, 220, 230, 170, 220, 230, 165, 205, 192, 210, 205, 200, 210, 185, 195, 202, 205, 195, 180, 200, 185, 240, 185, 220, 205, 205, 180, 201, 190, 208, 240, 180, 230, 195, 215, 190, 195, 215, 215, 220, 220, 230, 195, 190, 195, 209, 204, 170, 185, 205, 175, 210, 190, 180, 180, 160, 235, 200, 210, 180, 190, 197, 203, 205, 170, 200, 250, 200, 220, 200, 190, 170, 190, 220, 215, 206, 215, 185, 235, 188, 230, 195, 168, 190, 160, 200, 200, 189, 180, 190, 200, 220, 187, 240, 190, 180, 185, 210, 220, 219, 190, 193, 175, 180, 215, 210, 200, 190, 185, 220, 170, 195, 205, 195, 210, 190, 190, 180, 220, 190, 186, 185, 190, 180, 190, 170, 210, 240, 220, 180, 210, 210, 195, 160, 180, 205, 200, 185, 245, 190, 210, 200, 200, 222, 215, 240, 170, 220, 156, 190, 202, 221, 200, 190, 210, 190, 200, 165, 190, 185, 230, 208, 209, 175, 180, 200, 205, 200, 250, 210, 230, 244, 202, 240, 200, 215, 177, 210, 170, 215, 217, 198, 200, 220, 170, 200, 230, 231, 183, 192, 167, 190, 180, 180, 215, 160, 205, 223, 175, 170, 190, 240, 175, 230, 223, 196, 167, 195, 190, 250, 190, 190, 190, 170, 160, 150, 225, 220, 209, 210, 176, 260, 195, 190, 184, 180, 195, 195, 219, 225, 212, 202, 185, 200, 209, 200, 195, 228, 210, 190, 212, 190, 218, 220, 190, 235, 210, 200, 188, 210, 235, 188, 215, 216, 220, 180, 185, 200, 210, 220, 185, 231, 210, 195, 200, 205, 200, 190, 250, 185, 180, 170, 180, 208, 235, 215, 244, 220, 185, 230, 190, 200, 180, 190, 196, 180, 230, 224, 160, 178, 205, 185, 210, 180, 190, 200, 257, 190, 220, 165, 205, 200, 208, 185, 215, 170, 235, 210, 170, 180, 170, 190, 150, 230, 203, 260, 246, 186, 210, 198, 210, 215, 180, 200, 245, 200, 192, 192, 200, 192, 205, 190, 186, 170, 197, 219, 200, 220, 207, 225, 207, 212, 225, 170, 190, 210, 230, 210, 200, 238, 234, 222, 200, 190, 170, 220, 223, 210, 215, 196, 175, 175, 189, 205, 210, 180, 180, 197, 220, 228, 190, 204, 165, 216, 220, 208, 210, 215, 195, 200, 215, 229, 240, 207, 205, 208, 185, 190, 170, 208, 225, 190, 225, 185, 180, 165, 240, 220, 212, 163, 215, 175, 205, 210, 205, 208, 215, 180, 200, 230, 211, 230, 190, 220, 180, 205, 190, 180, 205, 190, 195]

In [ ]:
len(heights_in),len(weights_lb)

In [ ]:
# Importing the numpy library 

import numpy as np

In [ ]:
# Convert the lists into NumPy arrays
heights_in=np.array(heights_in)
weights_lb=np.array(weights_lb)

In [ ]:
heights_m=heights_in*0.0254
heights_m

In [ ]:
weight_kg=weights_lb*0.453592
weight_kg

In [ ]:
# Print the type of the created arrays
type(weights_lb)

In [ ]:
# Print the length of the created arrays

weights_lb.size

In [ ]:
# Convert the units of height and weight using appropriate factors (look in the video for the factors)

heights_m = heights_in * 0.0254           # 1 inch = 0.025 meters        1 inch = 2.54 cm
weights_kg = weights_lb * 0.45            # 1 lb = 0.45 Kgs

In [ ]:
# Print the newly created arrays

print(heights_m)
print(weights_kg)

Deriving new data from existing

Now, let us try to calculate the Body Mass Index (BMI) value for the players. The formula for BMI is:

BMI=Weight of the individual(Height of the individual)2

In [ ]:
# Calculate the bmi value based on the formula above

bmi = weights_kg/(heights_m**2)

In [ ]:
# Check the newly created array 'bmi'

bmi

In [ ]:
cond_bmi=bmi>25

In [ ]:
1==1

In [ ]:
bmi[:10]

In [ ]:
len(bmi[bmi>25])

In [ ]:
len(bmi)

# Indexing through arrays

For one-dimensional arrays, indexing is similar to Python lists - indexing starts at 0.

In [ ]:
# Obtain the 5th element from the array 'bmi'

bmi[4]

In [ ]:
# Obtain the 2nd last element from the array 'bmi'

bmi[-2]


In [ ]:
bmi[1013]

In [ ]:
# Obtain the first five elements from the array 'bmi'

bmi[:5]

In [ ]:
# Obtain the last three elements from the array 'bmi'
bmi[-3:]

In [ ]:
# Check for the elements where bmi value is less than 21

bmi<21

In [ ]:
# Filter the elements where bmi value is less than 21
bmi[bmi<21]

In [ ]:
n=10
lst=[i for i in range(n)]

In [ ]:
len(lst)

In [ ]:
cond_bmi=bmi<21

In [ ]:
bmi[cond_bmi]

In [ ]:
bmi.size

In [ ]:
bmi[:50].size

In [ ]:
# Count the number of elements where bmi value is less than 21

bmi[bmi<21].size

# Statistical Analysis
Statistical analysis includes finding minimum (.min()), maximum (max()), mean(.mean()), median(.median()), mode, correlation, covariance(.cov()), standard deviation (.std()) and variance (.var()) of the numpy array.

In [89]:
import random
arr=np.array([random.randrange(50,90) for i in range(50)])
arr

array([80, 84, 78, 88, 82, 70, 52, 53, 89, 73, 51, 60, 76, 72, 53, 50, 59,
       82, 72, 78, 71, 67, 59, 67, 69, 52, 52, 58, 58, 81, 86, 61, 51, 58,
       71, 52, 84, 63, 58, 87, 60, 85, 83, 50, 64, 84, 68, 77, 63, 58])

    max()
    min()
    mean()
    median()
    std()
    var()
    cov()
    corref()

In [ ]:
bmi.size

In [ ]:
bmi.max()

In [ ]:
bmi.min()

In [ ]:
bmi.max()-bmi.min()

In [ ]:
bmi.mean()

In [ ]:
dir(np)

In [ ]:
bmi[bmi<21].max()

In [ ]:
# Find the maximum bmi values among the players

bmi.max()

In [ ]:
# Find the minimum bmi values among the players

bmi.min()

In [ ]:
# Find the average bmi among the players

bmi.mean()

In [ ]:
np.median(bmi)

In [80]:
np.corrcoef(weight_kg)

1.0

In [81]:
# to find mode in numpy array
# Note: This step works fine in Jupyter Notebbok
import statistics as st

In [83]:
st.mode(bmi)

26.177571589521115

In [84]:
#Find the standard deviation of the bmi
bmi.std()

2.26658142199571

In [85]:
bmi.var()

5.137391342536095

In [86]:
#Find the variance of the bmi
np.var(bmi)

5.137391342536095

In [87]:
#Find the covariance of the bmi
np.cov(bmi)

array(5.1424578)

In [88]:
#Find the correlation of the bmi
np.corrcoef(bmi)

1.0

# Aggregation of array elements

    sum
    count
    max
    min
    average
    running total

In [90]:
# Reducing the elements to a single value based on addition
x=np.array([1,2,3,4,70,6])
print(np.add.reduce(x))

print(x.sum())

86
86


In [92]:
# Reducing the elements to a single value based on addition
y=np.array([[1,2,3,4,70,6],[1,2,3,4,70,6],[1,2,33,4,70,6]])
print(np.add.reduce(y))

print(y.sum())

[  3   6  39  12 210  18]
288


In [93]:
# Cummulative addition of elements at each element
print(np.add.accumulate(x))

[ 1  3  6 10 80 86]


In [ ]:
# Cummulative addition of elements at each element
print(np.add.accumulate(y))

In [94]:
print(np.cumsum(x))

[ 1  3  6 10 80 86]


In [95]:
print(np.cumsum(y))

[  1   3   6  10  80  86  87  89  92  96 166 172 173 175 208 212 282 288]


# Practice Exercise 1

You are provided with 2 lists that contain the data of an ecommerce website. The first list contains the data for the number of items sold for a particular product and the second list contains the price of the product sold. As a part of this exercise, solve the questions that are provided below.

In [97]:
number = [8, 9, 9, 1, 6, 9, 5, 7, 3, 9, 7, 3, 4, 8, 3, 5, 8, 4, 8, 7, 5, 7, 3, 6, 1, 2, 7, 4, 7, 7, 8, 4, 3, 4, 2, 2, 2, 7, 3, 5, 6, 1, 1, 3, 2, 1, 1, 7, 7, 1, 4, 4, 5, 6, 1, 2, 7, 4, 5, 8, 1, 4, 8, 6, 2, 4, 3, 7, 3, 6, 2, 3, 3, 3, 2, 4, 6, 8, 9, 3, 9, 3, 1, 8, 6, 6, 3, 3, 9, 4, 6, 4, 9, 6, 7, 1, 2, 8, 7, 8, 1, 4]

In [98]:
price = [195, 225, 150, 150, 90, 60, 75, 255, 270, 225, 135, 195, 30, 15, 210, 105, 15, 30, 180, 60, 165, 60, 45, 225, 180, 90, 30, 210, 150, 15, 270, 60, 210, 180, 60, 225, 150, 150, 120, 195, 75, 240, 60, 45, 30, 180, 240, 285, 135, 165, 180, 240, 60, 105, 165, 240, 120, 45, 120, 165, 285, 225, 90, 105, 225, 45, 45, 45, 75, 180, 90, 240, 30, 30, 60, 135, 180, 15, 255, 180, 270, 135, 105, 135, 210, 180, 135, 195, 225, 75, 225, 15, 240, 60, 15, 180, 255, 90, 15, 150, 230, 150]

In [100]:
quantity=np.array(number)
unit_price=np.array(price)
sales=quantity*unit_price

# How many different products are sold by the company in total?

* 99
* 100
* 101
* 102

In [101]:
quantity.size

102

In [99]:
# Type your code here
import numpy as np
number=np.array(number)

number.size


102

# How many items were sold in total?

* 460
* 490
* 500
* 520

In [102]:
quantity.sum()

490

In [ ]:
# Type your code here
number.sum()

In [103]:
unit_price.mean()

139.01960784313727

# What is the average price of the products sold by the ecommerce company?

* 139
* 151
* 142
* 128

In [ ]:
# Type your code here

price=np.array(price)

price.mean()

# What is the price of the costliest item sold?

* 225
* 310
* 280
* 285

In [ ]:
# Type your code here
price.max()

In [104]:
sales.sum()

67100

# What is the total revenue of the company? [Revenue = Price*Quantity]

* 67100
* 53900
* 45300
* 71200

In [ ]:
# Type your code here

Revenue= (price*number).sum()
# Revenue=Revenue.sum()
Revenue

In [105]:
quantity[19]

7

In [106]:
quantity[49]

1

In [107]:
if quantity[19]>quantity[49]:
    print(True)
else:
    print(False)

True


# Demand for the 20th product in the list is more than the 50th product. [True/False]

* True
* False
* Can't be calculated

In [108]:
# Type your code here
number[19]>number[49]

True

In [ ]:
number[19]

In [ ]:
number[49]

    How many products fall under the category of expensive goods? An expensive good is that good whose price is more than the average price of the products sold by the company.

    * 48
    * 50
    * 52
    * 54

In [ ]:
# Type your code here

price[price > price.mean()].size 

# Linear Algebra

    7x+5y-3z=16
    3x-5y+2z=-8
    5x+3y-7z=0

In [1]:
import numpy as np

In [3]:
a = np.array([ [7,5,-3],[3,-5,2],[5,3,-7] ])
a

array([[ 7,  5, -3],
       [ 3, -5,  2],
       [ 5,  3, -7]])

In [4]:
b = np.array([16,-8,0])
b

array([16, -8,  0])

In [7]:
# to get the solution of the equation use np.linalg.solve()

a = np.array([ [7,5,-3],[3,-5,2],[5,3,-7] ])
b = np.array([16,-8,0])

np.linalg.solve(a,b)

array([1., 3., 2.])

In [8]:
# to get the inverse of the matrix use np.linalg.inv()
a = np.array([ [7,5,-3],[3,-5,2],[5,3,-7] ])
print(a)

np.linalg.inv(a)

[[ 7  5 -3]
 [ 3 -5  2]
 [ 5  3 -7]]


array([[ 0.11328125,  0.1015625 , -0.01953125],
       [ 0.12109375, -0.1328125 , -0.08984375],
       [ 0.1328125 ,  0.015625  , -0.1953125 ]])

In [ ]:
np.round(np.linalg.inv(a),2)

In [ ]:
# to get the determinant of the matrix use np.linalg.det()
a = np.array([ [7,5,-3],[3,-5,2],[5,3,-7] ])
print(a)

np.linalg.det(a)

In [9]:
A = np.array([[1, 1],
              [0, 1]])

B = np.array([[2, 0],
              [3, 4]])

    mxn
    pxq
    n=p

    4x3
    3x2

In [10]:
A,B

(array([[1, 1],
        [0, 1]]),
 array([[2, 0],
        [3, 4]]))

In [11]:
A * B     # elementwise product

array([[2, 0],
       [0, 4]])

In [12]:
A @ B     # matrix product or dot product

array([[5, 4],
       [3, 4]])

In [ ]:
A.dot(B)  # another matrix product

In [13]:
p = [4, 2]
q = [5, 6]
product = np.cross(p,q)       # cross product
print(product)

14


# Creating Special NumPy Arrays
There are multiple ways to create numpy arrays, the most commmon ones being:

Convert lists or tuples to arrays using np.array()
Initialise arrays of fixed size (when the size is known)

The following ways are commonly used:

* np.arange(): Create array with increments of a fixed step size
* np.ones(): Create array of 1s
* np.zeros(): Create array of 0s
* np.linspace(): Create array of fixed length
* np.full(): Create a constant array of any number ‘n’
* np.tile(): Create a new array by repeating an existing array for a particular number of times
* np.eye(): Create an identity matrix of any dimension
* np.random.random(): Create array of random numbers between 0 and 1
* np.random.randint(): Create a random array of integers within a particular range

In [14]:
# importing the NumPy library
import numpy as np

In [15]:
# creating an array with 4 elements and increments of a fixed step size
# default step size is one
np_1 = np.arange(4)
np_1

array([0, 1, 2, 3])

In [16]:
# creating an array between 2 and 11 elements and increments of a fixed step size 2
# default step size is one
np_1 = np.arange(2,11,2)
print(np_1)

[ 2  4  6  8 10]


In [ ]:
# creating an array of zeroes with 5 elements


In [ ]:
#answer here

In [18]:
# specifying the data type of the array in the provided attribute

np_2 = np.zeros(5,dtype="int")
np_2.dtype, np_2

(dtype('int32'), array([0, 0, 0, 0, 0]))

    I want array of zeros with 12 elements having shape 3x4

In [23]:
np_3=np.zeros((3,2,2),dtype='int')
np_3

array([[[0, 0],
        [0, 0]],

       [[0, 0],
        [0, 0]],

       [[0, 0],
        [0, 0]]])

In [ ]:
np_2

In [ ]:
# adding the two arrays: np_1 and np_2
print(np_1 + np_2)

In [ ]:
# check the dimension of the arrays
# array np_1
np_1.size

In [ ]:
# check the dimension of the arrays
# array np_2
np_2.size

In [29]:
np_3 = np.ones(12,dtype='int')
np_3

array([[1, 1, 1, 1],
       [1, 1, 1, 1],
       [1, 1, 1, 1]])

In [26]:
# creating a third array with all the 5 elements as ones

np_3 = np.ones((4,2),dtype='int')
print(np_3)

[[1 1]
 [1 1]
 [1 1]
 [1 1]]


In [ ]:
# check the dimension of the arrays
np_3.size

In [ ]:
# creating a third array with all the 5 elements as ones

np_4 = np.ones(5,dtype=int)

# checking the created array:np_3
print(np_4)

In [ ]:
# subtracting the array np_3 from array np_2

print(np_2 - np_3)

# np.linespace()

In [31]:
np_5=np.linspace(5,10,2)
np_5

array([ 5., 10.])

In [33]:
# creating an array of fixed length
np_5 = np.linspace(1,90,5)

# checking the created array:np_4
print(np_5)

[ 1.   23.25 45.5  67.75 90.  ]


In [34]:
# creating an array of fixed length
np_6 = np.linspace(1,10,10)

# checking the created array:np_4
print(np_6)

[ 1.  2.  3.  4.  5.  6.  7.  8.  9. 10.]


In [35]:
np_7=np.linspace(1,90,9,retstep=True)
print(np_7)

(array([ 1.   , 12.125, 23.25 , 34.375, 45.5  , 56.625, 67.75 , 78.875,
       90.   ]), 11.125)


1
12.125

11.125

23.25

In [38]:
np.linspace(1,10,10)

array([ 1.,  2.,  3.,  4.,  5.,  6.,  7.,  8.,  9., 10.])

    np.zeros  it will return array of 0s only
    np.ones   it will return array of 1s only
    np.full   It will return array of given interger number

In [43]:
np.full(5,45)

array([45, 45, 45, 45, 45])

In [ ]:
# Creating a 4 x 3 array of 7s using np.full()
# The default data type here is 'int' only
np.full((4,3), 7)

In [44]:
# Create a 3 x 3 identity matrix using np.eye()
# The default data type here is float. So if we want integer values, we need to specify the dtype to be int
np.eye(3, dtype = int)

array([[1, 0, 0],
       [0, 1, 0],
       [0, 0, 1]])

In [48]:
np.eye(5, dtype = int)

array([[1, 0, 0, 0, 0],
       [0, 1, 0, 0, 0],
       [0, 0, 1, 0, 0],
       [0, 0, 0, 1, 0],
       [0, 0, 0, 0, 1]])

In [46]:
np.eye(2)

array([[1., 0.],
       [0., 1.]])

In [52]:
# Create a 3 x 3 array of random numbers between 0 and 1
np.random.random([3, 4])*10

array([[9.34237517, 7.33951301, 2.49641023, 6.92271129],
       [8.83879972, 6.46965134, 7.01394776, 0.67493281],
       [9.02277405, 9.57233162, 8.16163586, 8.30723137]])

.randint()

random.randint(low, high=None, size=None, dtype=int)

Return random integers from low (inclusive) to high (exclusive).

In [54]:
# Create a 4 x 4 random array of integers ranging from 0 to 9
np.random.randint(20, 30, (4,5))

array([[24, 23, 24, 28, 29],
       [27, 20, 26, 21, 29],
       [28, 27, 28, 25, 23],
       [25, 29, 28, 25, 26]])

In [55]:
# Create a 4 x 4 random array of integers ranging from 0 to 9
np.random.randint(0, 10, (4,4))

array([[0, 4, 6, 6],
       [0, 6, 4, 0],
       [7, 7, 6, 0],
       [9, 9, 9, 6]])